tests/test_engine.py

```diff
@@ -1331,6 +1331,56 @@ def my_task(x):
         assert state.is_failed()
         assert mock.call_count == 2
 
+    async def test_retry_condition_fn_no_retries_after_failure(
+        self, mock_anyio_sleep, prefect_client, flow_run, result_factory
+    ):
+        # the flow run must be running prior to running tasks
+        await prefect_client.set_flow_run_state(
+            flow_run_id=flow_run.id,
+            state=Running(),
+        )
+
+        # Mocks to check call counts
+        mock = MagicMock()
+        mock_2 = MagicMock()
+
+        # Do not retry
+        def is_retriable(task, task_run, state):
+            mock_2()
+            return True is False
+
+        # Can retry more than once
+        @task(retries=5, retry_condition_fn=is_retriable)
+        def my_task(x):
+            mock(x)
+            raise ValueError("try again, but only once")
+
+        # Create a task run to test
+        task_run = await prefect_client.create_task_run(
+            task=my_task,
+            flow_run_id=flow_run.id,
+            state=Pending(),
+            dynamic_key="0",
+        )
+
+        # Actually run the task
+        state = await orchestrate_task_run(
+            task=my_task,
+            task_run=task_run,
+            wait_for=None,
+            parameters={"x": quote(1)},
+            result_factory=result_factory,
+            interruptible=False,
+            client=prefect_client,
+            log_prints=False,
+        )
+
+        # Check that the task failed after only one attempt
+        assert state.is_failed()
+        assert mock.call_count == 1
+        # Check that the retry condition function was only called once
+        assert mock_2.call_count == 1
```

Comment:
I think the call count should be 0 and

`# Check that the retry condition function was only called once`

is incorrect here.

Reply:
No, the `retry_condition_fn` is always called on a fatal terminal state. It runs before a fail state is proposed to the server and the server adjudicates whether to fail or retry. We expect it to be called once.

Reply:
Ah yeah I see. I'd forgotten that. Hm okay.